Update Readme (#28)
virtee/snpguest: README.md

Reviewer: In addition to the other comments I left, I would also add a section on how to compile the tool. You also fail to mention the whole hyper-v section.

# snpguest

`snpguest` is a Command Line Interface utility for managing an AMD SEV-SNP enabled guest. This tool allows users to interact with the AMD SEV-SNP guest firmware device enabling various operations such as: attestation, certificate management, derived key fetching, and more.

- [Usage](#usage)
  - [1. help](#1-help)
  - [2. certificates](#2-certificates)
  - [3. display](#3-display)
  - [4. fetch](#4.fetch)
  - [5. key](#5-key)
  - [6. report](#6-report)
  - [7. verify](#7-verify)

- [Extended Attestation Workflow](#extended-attestation-workflow)
- [Regular Attestation Workflow](#regular-attestation-workflow)
- [Global Options](#global-options)
- [Extended Attestation Flowchart](#extended-attestation-flowchart)
- [Regular Attestation Flowchart](#regular-attestation-flowchart)
- [Building](#building)
- [Reporting Bugs](#reporting-bugs)

## Usage
### 1. `help`

Every `snpguest` (sub)command comes with a `--help` option for a description on its use. 

Usage
```console
snpguest --help
```

> Apart from `help`, there are 6 other subcommands for `snpguest` that have been listed below with their purpose.

Reviewer: ```suggestion

```

Example
```console
snpguest certificates pem ./certs
```

### 3. `display` 
Displays files in human readable form. 

Usage
```console
snpguest display <SUBCOMMAND>
```
Subcommands

1. `report`

    When used for displaying a report, it prints the attestation report contents into the terminal. The user has to provide the path of the stored attestation report to display.

    Usage
    ```bash
    snpguest display report $ATT_REPORT_PATH
    ```

    Argument

    - `$ATT_REPORT_PATH` : Specifies the path of the stored attestation report to display.

    Example
    ```bash
    snpguest display report attestation-report.bin
    ```

2. `key`

    When used for displaying the fetched derived key's contents, it prints the derived key in hex format into the terminal. The user has to provide the path of a stored derived key to display.

    Usage

    ```bash
    snpguest display key $KEY_PATH
    ```
    Argument

    - `$KEY_PATH` : Specifies the path of the stored derived key to display.

    Example
    ```bash
    snpguest display key derived-key.bin
    ```
### 4. `fetch`

> Command to Requests certificates.

Reviewer: ```suggestion
Command to Requests certificates from the KDS.
```

Usage
```bash
snpguest fetch <SUBCOMMAND>
```

Subcommands
1. `ca`

>     Requests the certificate authority chain (ARK & ASK) from the KDS. The user needs to specify the certificate encoding to store the certificates in (PEM or DER). Currently, only PEM and DER encodings are supported. Both certificates will be in the same encoding. The user must specify their host processor model (Milan or Genoa). The user also needs to provide the path to the directory where the certificates will be stored. If the certificates already exist in the provided directory, they will be overwritten.

Reviewer: ```suggestion
    Requests the certificate authority chain (ARK & ASK) from the KDS. The user needs to specify the certificate encoding to store the certificates in (PEM or DER). Currently, only PEM and DER encodings are supported. Both certificates will be in the same encoding. The user must specify their host processor model. The user also needs to provide the path to the directory where the certificates will be stored. If the certificates already exist in the provided directory, they will be overwritten.
```

    Usage
    ```bash
    snpguest fetch ca $ENCODING $PROCESSOR_MODEL $CERTS_DIR
    ```
    Arguments
    
    - `$ENCODING` : Specifies the certificate encoding to store the certificates in (PEM or DER)

>     - `$PROCESSOR_MODEL` : Specifies the host processor model (Milan or Genoa). 

Reviewer: ```suggestion
    - `$PROCESSOR_MODEL` : Specifies the host processor model. 
```

    
    - `$CERTS_DIR` : Specifies the directory to store the certificates in.

    Example
    ```bash
    snpguest fetch ca der milan ./certs-kds
    ```

2. `vcek`

>     Requests the VCEK certificate from the KDS. The user needs to specify the certificate encoding to store the certificate in (PEM or DER). Currently, only PEM and DER encodings are supported. The user must specify their host processor model (Milan or Genoa). The user also needs to provide the path to the directory where the VCEK will be stored and the path to a stored attestation report that will be used to request the VCEK. If the certificate already exists in the provided directory, it will be overwritten.

Reviewer: ```suggestion
    Requests the VCEK certificate from the KDS. The user needs to specify the certificate encoding to store the certificate in (PEM or DER). Currently, only PEM and DER encodings are supported. The user must specify their host processor model. The user also needs to provide the path to the directory where the VCEK will be stored and the path to a stored attestation report that will be used to request the VCEK. If the certificate already exists in the provided directory, it will be overwritten.
```

    Usage
    ```bash
    snpguest fetch vcek $ENCODING $PROCESSOR_MODEL $CERTS_DIR $ATT_REPORT_PATH
    ```
    Arguments
    
    - `$ENCODING` : Specifies the certificate encoding to store the certificates in (PEM or DER)

>     - `$PROCESSOR_MODEL` : Specifies the host processor model (Milan or Genoa). 

Reviewer: ```suggestion
    - `$PROCESSOR_MODEL` : Specifies the host processor model. 
```

    
    - `$CERTS_DIR` : Specifies the directory to store the certificates in. 

    - `$ATT_REPORT_PATH` : Specifies the path of the stored attestation report.

    Example
    ```bash
    snpguest fetch vcek der milan ./certs-kds attestation-report.bin
    ```

### 5. `key` 

> Creates the derived key based on input parameters and stores it. `$KEY_PATH` is the path to store the derived key. `$ROOT_KEY_SELECT` is the root key from which to derive the key (either "vcek" or "vmrk"). The `--guest_field_select` option specifies which Guest Field Select bits to enable as a 6-digit binary string. For each bit, 0 denotes off, and 1 denotes on (e.g., `--guest_field_select 100001` denotes Guest Policy:On, Image ID:Off, Family ID:Off, Measurement:Off, SVN:Off, TCB Version:On). The `--guest_svn` option specifies the guest SVN to mix into the key, and the `--tcb_version` option specifies the TCB version to mix into the derived key. The `--vmpl` option specifies the VMPL level the Guest is running on and defaults to 1.

Reviewer: On guest field select explanation, instead of showing example, just explain what each bit is. User can see example underneath.

```bash
# Creating and storing a derived key
snpguest key derived-key.bin vcek --guest_field_select 100001 --guest_svn 2 --tcb_version 1 --vmpl 3
```

### 6. `report` 

Requests an attestation report from the host and writes it to a file with the provided request data and VMPL. The attestation report is written in binary format to the specified report path. The user can pass 64 bytes of data in any file format into `$REQUEST_FILE` to request the attestation report. The `--random` flag can be used to generate and use random data for the request, and for Microsoft Hyper-V guests, the `--platform` flag is required to use pre-generated request data. The data already generated from the hypervisor will be written into the file provided in `$REQUEST_FILE`. `VMPL` is an optional parameter that defaults to 1. `--random` is not available in Hyper-V, as the request data is pre-generated.

Usage
```bash
snpguest report $ATT_REPORT_PATH $REQUEST_FILE [-v, --vmpl] [-r, --random] [-p, --platform]
```

Arguments

- `$ATT_REPORT_PATH` : Specifies the path where the attestation report would be stored.

- `$REQUEST_FILE` : File where the data generated from the hypervisor will be written into.

> Options

Reviewer: Missing platform flag

# Creating and storing a derived key
snpguest key derived-key.bin vcek --guest_field_select 100001 --guest_svn 2 --tcb_version 1 --vmpl 3
```

### 6. `report` 

Requests an attestation report from the host and writes it to a file with the provided request data and VMPL. The attestation report is written in binary format to the specified report path. The user can pass 64 bytes of data in any file format into `$REQUEST_FILE` to request the attestation report. The `--random` flag can be used to generate and use random data for the request, and for Microsoft Hyper-V guests, the `--platform` flag is required to use pre-generated request data. The data already generated from the hypervisor will be written into the file provided in `$REQUEST_FILE`. `VMPL` is an optional parameter that defaults to 1. `--random` is not available in Hyper-V, as the request data is pre-generated.

Usage
```bash
snpguest report $ATT_REPORT_PATH $REQUEST_FILE [-v, --vmpl] [-r, --random] [-p, --platform]
```

Arguments

- `$ATT_REPORT_PATH` : Specifies the path where the attestation report would be stored.

- `$REQUEST_FILE` : File where the data generated from the hypervisor will be written into.

Options

> - `-r, --random`: Generate 64 random bytes of data for the report request. (Not used in Hyper-V)

Reviewer 2: ```suggestion
- `-r, --random`: Generate 64 random bytes of data for the report request (not available for Hyper-V).
```

> - `-p, --platform` : Use platform provided 64 bytes of data for the request report. 

Reviewer 2: ```suggestion
- `-p, --platform` : Use platform provided 64 bytes of data for the request report (only available for Hyper-V).
```

> `snpguest` is a Command Line Interface utility for managing the AMD SEV-SNP guest environment. This tool allows users to interact with the AMD SEV-SNP guest firmware device enabling various operations related to attestation, certificate management, key fetching, and more.

Author: I believe @tylerfanelli had suggested to use the term fetch instead of derivation. Can we come to a consensus?

> `snpguest` is a Command Line Interface utility for managing the AMD SEV-SNP guest environment. This tool allows users to interact with the AMD SEV-SNP guest firmware device enabling various operations related to attestation, certificate management, key fetching, and more.

Reviewer: Derived key fetching

   - The `display_derived_key` function reads a derived key from the specified file path, formats it as a hexadecimal string, and prints it to the console.
   - The key is displayed in a human-readable format with 16 bytes per line.
   - If the `quiet` flag is not set, the derived key is printed.

Fetch.rs:  
This code defines a CLI for fetching certificates from a Key Distribution Service (KDS). The code aims to provide a convenient CLI interface for users to fetch certificates from a KDS, supporting different processor models and certificate formats. It also handles directory creation and certificate writing. It includes the following functionality:

1. **Command-Line Interface (CLI) Definition**: The code defines a CLI with two subcommands:

- `CA`: Fetches the certificate authority chain (ARK & ASK) from the KDS.

- `Vcek`: Fetches the VCEK (Virtualization Code Execution Key) from the KDS.

2. **Enumeration and Parsing**: It defines an enumeration called `ProcType` to represent processor types ("Milan" or "Genoa"). It also implements the `FromStr` trait for parsing processor types and the `fmt::Display` trait for displaying them.

> 3. **Command Handling**: The `cmd` function is responsible for handling the selected subcommand and performing the corresponding action. It calls either the `cert_authority::fetch_ca` function or the `vcek::fetch_vcek` function based on the selected subcommand.

Reviewer: Same thing as above. Very good summary, but not part of a README

## Usage
#### [Extended Attestation Workflow](#extended-attestation-flowchart)

**Step 1.**

```bash
snpguest report $ATT_REPORT_PATH $REQUEST_FILE [-v, --vmpl] [-r, --random] [-p, --platform]
```

**Step 2.**

```bash
snpguest certificates $ENCODING $CERTS_DIR
```

**Step 3.**

```bash
snpguest verify certs $CERTS_DIR
```

**Step 4.**

```bash
snpguest verify attestation $CERTS_DIR $ATT_REPORT_PATH [-t, --tcb] [-s, --signature]
```

#### [Regular Attestation Workflow](#regular-attestation-flowchart)

**Step 1.**

```bash
snpguest report $ATT_REPORT_PATH $REQUEST_FILE [-v, --vmpl] [-r, --random] [-p, --platform]
```

**Step 2.**

```bash
snpguest fetch ca $ENCODING $PROCESSOR_MODEL $CERTS_DIR
```

**Step 3.**

```bash
snpguest fetch vcek $ENCODING $PROCESSOR_MODEL $CERTS_DIR $ATT_REPORT_PATH
```

**Step 4.**

```bash
snpguest verify certs $CERTS_DIR
```

**Step 5.**

```bash
snpguest verify attestation $CERTS_DIR $ATT_REPORT_PATH [-t, --tcb] [-s, --signature]

> ```

Reviewer: Explain what each step is doing.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Requesting ARK & ASK certificates from KDS
snpguest fetch ca der milan ./certs-kds
```

#### snpguest fetch vcek

```bash
snpguest fetch vcek $ENCODING $PROCESSOR_MODEL $CERTS_DIR $ATT_REPORT_PATH
```

Requests the VCEK certificate from the KDS. The user needs to specify the certificate encoding to store the certificate in (PEM or DER). Currently, only PEM and DER encodings are supported. The user must specify their host processor model (Milan or Genoa). The user also needs to provide the path to the directory where the VCEK will be stored and the path to a stored attestation report that will be used to request the VCEK. If the certificate already exists in the provided directory, it will be overwritten.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Requesting VCEK
snpguest fetch vcek der milan ./certs-kds attestation-report.bin
```

> #### snpguest guest verify certs

Reviewer: ```suggestion
#### snpguest verify certs
```

```bash
snpguest verify certs $CERTS_DIR
```

Verifies that the provided certificate chain has been properly signed by each certificate. The user needs to provide a directory where all three certificates (ARK, ASK, and VCEK) are stored. An error will be raised if any of the certificates fail verification.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Verifying Certs
snpguest verify certs ./certs
```

> #### snpguest guest verify attestation

Reviewer: ```suggestion
#### snpguest guest attestation
```

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Creating and storing a derived key
snpguest key derived-key.bin vcek --guest_field_select 100001 --guest_svn 2 --tcb_version 1 --vmpl 3
```

#### snpguest guest display report

```bash
snpguest display report $ATT_REPORT_PATH
```

Prints the attestation report contents into the terminal. The user has to provide a path to a stored attestation report to display.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Displaying an attestation report
snpguest display report attestation-report.bin
```

#### snpguest guest display key

```bash
snpguest display key $KEY_PATH
```

Prints the derived key contents in hex format into the terminal. The user has to provide the path of a stored derived key to display.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Displaying a derived key
snpguest display key derived-key.bin
```

> ---

Reviewer: This should be the first part of the README. The way I would divide would be:
A section for the main command, and then a small explanation of the subcommands. 
For Example:
```
### verify
Command to do different types of verification
#### verify certs

snpguest verify certs $CERTS_DIR

Verifies that the provided certificate chain has been properly signed by each certificate. The user needs to provide a directory where all three certificates (ARK, ASK, and VCEK) are stored. An error will be raised if any of the certificates fail verification.

Options:
- `-h, --help`: Show a help message.

Example usage:
# Verifying Certs
snpguest verify certs ./certs
```

Or something along those lines. Again check SEVCTL README

- `<attestation-report-path>`: Specify the path to an attestation report to be used to request the VCEK.

Example:

```bash
snpguest fetch vcek der milan ./certs-kds attestation-report.bin
```

## Verifying Attestation

### Verifying certificate chain:

```bash
snpguest verify certs <certs-directory>
```

- `<certs-directory>`: Specify the path to a directory where the certificates are stored. A full certificate chain is required (ASK, ARK, and VCEK).

Example:

```bash
snpguest verify certs ./certs
```

### Verifying Attestation Report:

```bash
snpguest verify attestation <certs-directory> <attestation-report-path> [-t, --tcb] [-s, --signature]
```

- `<certs-directory>`: Specify the path to a directory where the certificates are stored.
- `<attestation-report-path>`: Specify the path to an attestation report to be verified.
- `-t, --tcb`: Just verify the TCB contents of the attestation report.
- `-s, --signature`: Just verify the signature of the attestation report.

Example:

```bash
# Verify Attestation
snpguest verify attestation ./certs attestation-report.bin
# Verify Attestation Signature only
snpguest verify attestation ./certs attestation-report.bin --signature

> ```

Reviewer: This seems very repetitive to what you have in the previous section. The examples should be way shorter. What I would do is go back to the attestation workflow examples and just explain what you're doing in each step. Maybe then add an example for the derived key.

> **SNPguest** is a Command Line Interface utility for managing the AMD SEV-SNP guest environment. This tool allows users to interact with the AMD SEV-SNP guest firmware device enabling various operations related to attestation, certificate management, key generation, and more.

Reviewer 3: ```suggestion
**SNPguest** is a Command Line Interface utility for managing the AMD SEV-SNP guest environment. This tool allows users to interact with the AMD SEV-SNP guest firmware device enabling various operations related to attestation, certificate management, key fetching, and more.
```

I don't think there is any key generation in this tool, moreso fetching keys from the KDS that already exist.

## Table of Contents

- [Summary](#summary)
- [File Structure](#file-structure)
- [Usage](#usage)
  - [Global Options](#global-options)
  - [Commands](#commands)
    - [snpguest report](#snpguest-report)
    - [snpguest certificates](#snpguest-certificates)
    - [snpguest fetch ca](#snpguest-fetch-ca)
    - [snpguest fetch vcek](#snpguest-fetch-vcek)
    - [snpguest guest verify certs](#snpguest-guest-verify-certs)
    - [snpguest guest verify attestation](#snpguest-guest-verify-attestation)
    - [snpguest key](#snpguest-key)
    - [snpguest guest display report](#snpguest-guest-display-report)
    - [snpguest guest display key](#snpguest-guest-display-key)
- [SNP Examples](#snp-examples)
- [Reporting Bugs](#reporting-bugs)
- [Extended Attestation Flowchart](#extended-attestation-flowchart)
- [Regular Attestation Flowchart](#regular-attestation-flowchart)

## Summary

The "src" directory in the `snpguest` GitHub repository contains several Rust source code files that collectively implement the functionality of the SNP Guest utility. Here's a summary of the functionality of each of these files:

1. **`certs.rs`**:

- This file contains code related to managing certificates.

> - It defines a structure for managing certificate paths (`CertPaths`) and functions for obtaining extended certificates from the AMD PSP.

Reviewer 2: ```suggestion
- It defines a structure for managing certificate paths (`CertPaths`) and functions for obtaining extended certificates from the AMD Secure Processor.
```

Reviewer 3: Just for my reference, what's the difference between the two? I usually use them interchangeably. Or is this change because some users may not know what PSP stands for?

2. **`display.rs`**:

- This file defines the CLI for displaying attestation reports and derived keys.

> - It contains subcommands for displaying attestation reports and derived keys.

Reviewer 2: I think it would be beneficial to actually list out each sub-command, and what they are expected to do and how they are expected to be used.

Reviewer 3: Right, the [sevctl README](https://github.com/virtee/sevctl/blob/main/README.md) does just this, and can be used for reference.

> - It contains subcommands for displaying attestation reports and derived keys.

Reviewer 2: I think it would be beneficial to actually list out each sub-command, and what they are expected to do and how they are expected to be used.

- Submodules `report_display` and `key_display` handle the respective display functionalities.

3. **`fetch.rs`**:

- This file defines the CLI for fetching certificates.

> - It contains subcommands for fetching various types of certificates from the AMD PSP.

Reviewer 2: ```suggestion
- It contains subcommands for fetching various types of certificates from the AMD Secure Processor.
```

## File Structure

Main.rs:
This code serves as a CLI utility for managing and interacting with the SNP Guest environment in virtualization scenarios. It offers various operations related to certificates, attestation reports, key generation, and verification, making it a versatile tool for SNP Guest management. The choice of specific operations is determined by the selected subcommand.

1. **Module Imports**:

- The code imports several modules that provide functionality for certificates, display, fetching data, key generation, reporting, and verification.

2. **Feature Flags**:

- Conditional compilation is used based on a feature flag named `hyperv`. Depending on whether this feature is enabled or not, it is used to support different virtualization platforms.

3. **Command-Line Interface (CLI)**:

- The main program defines a CLI interface using the `structopt` crate.

- It has subcommands for different operations, including reporting, certificate management, fetching, verification, displaying, and key generation.

- The `quiet` flag is provided to suppress console output if set.

4. **Main Function** (`main`):

> - The main function initializes the `env_logger` for logging.

Reviewer 2: It would be good to explain how to set the environment variable to change the log level at run-time.

4. **Certificate Chain Verification** (Inside `certificate_chain` module):

- The `validate_cc` function verifies a certificate chain consisting of ARK (AMD Root Key), ASK (AMD Signing Key), and VCEK (Virtualization Code Encryption Key).

- It reads these certificates from a specified directory, validates their signatures, and ensures the certificates are properly chained.

- The result of each verification step is printed to the console unless the `quiet` flag is set.

5. **Attestation Report Verification** (Inside `attestation` module):

- The `verify_attesation` function verifies an attestation report's contents.

- It reads an attestation report from a file and compares the reported values with those in the VCEK certificate extensions.

- The verification includes checking Boot Loader, TEE (Trusted Execution Environment), SNP (Secure Nested Paging), Microcode, and Hardware ID values.

- The function also allows for separate verification of the TCB (Trusted Computing Base) and the digital signature.

- Verification results are printed to the console.

6. **Helper Functions**:

> - Several helper functions are defined within the modules to facilitate various tasks, including locating certificates in a directory, comparing certificate values, and verifying attestation report signatures.

Reviewer: Although this is a good summary of how the code works and I appreciate the effort, I don't think this is needed in the README. The README should focus on how to use the tool, not so much on how it works. Unless you disagree @larrydewey @tylerfanelli

Reviewer 2: I agree

Reviewer 3: I agree as well. However, much of the code comments would fit nicely in their respective source code files.

Reviewer 3: Much of this would also be useful in a man page as well.

## Usage

### Global Options

- **-q, --quiet**: Suppress console output.

### Commands

#### snpguest report

```bash
snpguest report $ATT_REPORT_PATH $REQUEST_FILE [-v, --vmpl] [-r, --random] [-p, --platform]
```

> Requests an attestation report from the host and writes it to a file with the provided request data and VMPL. The attestation report is written in binary format to the specified report path. The user can pass 64 bytes of data in any file format into `$REQUEST_FILE` to request the attestation report. The `--random` flag can be used to generate and use random data for the request, and for Microsoft Hyper-V guests, the `--platform` flag is required to use pre-generated request data. If the `--random` flag is used, the data will be written into the file provided in `$REQUEST_FILE`. `VMPL` is an optional parameter that defaults to 1.

Reviewer 3: ```suggestion
Requests an attestation report from the host and writes it to a file with the provided request data and VMPL. The attestation report is written in binary format to the specified report path. The user can pass 64 bytes of data in any file format into `$REQUEST_FILE` to request the attestation report. The `--random` flag can be used to generate and use random data for the request.

For Microsoft Hyper-V guests, the `--platform` flag is required to use pre-generated request data. The data already generated from the hypervisor will be written into the file provided in `$REQUEST_FILE`. `VMPL` is an optional parameter that defaults to 1. `--random` is not available in Hyper-V, as the request data is pre-generated.
```

Options:
- `-h, --help`: Show a help message.

> - `-r, --random`: Generate 64 random bytes of data for the report request.

Reviewer 3: ```suggestion
- `-r, --random`: Generate 64 random bytes of data for the report request. (Not used in Hyper-V)
```

```

Requests the VCEK certificate from the KDS. The user needs to specify the certificate encoding to store the certificate in (PEM or DER). Currently, only PEM and DER encodings are supported. The user must specify their host processor model (Milan or Genoa). The user also needs to provide the path to the directory where the VCEK will be stored and the path to a stored attestation report that will be used to request the VCEK. If the certificate already exists in the provided directory, it will be overwritten.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Requesting VCEK
snpguest fetch vcek der milan ./certs-kds attestation-report.bin
```

#### snpguest guest verify certs

```bash
snpguest verify certs $CERTS_DIR
```

Verifies that the provided certificate chain has been properly signed by each certificate. The user needs to provide a directory where all three certificates (ARK, ASK, and VCEK) are stored. An error will be raised if any of the certificates fail verification.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Verifying Certs
snpguest verify certs ./certs
```

#### snpguest guest verify attestation

```bash

> snpguest verify attestation $CERTS_DIR $ATT_REPORT_PATH [-t, --tcb] [-s, --signature]

Reviewer 3: Why is `[-s, --signature]` an option? Shouldn't the signature be a part of the attestation report?

@DGonzalezVillal @larrydewey as in, must there be a separate flag for the signature?

Reviewer: Hey, this allows you to perform attestation on only the signature part or only the tcb part.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Creating and storing a derived key
snpguest key derived-key.bin vcek --guest_field_select 100001 --guest_svn 2 --tcb_version 1 --vmpl 3
```

#### snpguest guest display report

```bash
snpguest display report $ATT_REPORT_PATH
```

Prints the attestation report contents into the terminal. The user has to provide a path to a stored attestation report to display.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Displaying an attestation report
snpguest display report attestation-report.bin
```

#### snpguest guest display key

```bash
snpguest display key $KEY_PATH
```

Prints the derived key contents in hex format into the terminal. The user has to provide the path of a stored derived key to display.

Options:
- `-h, --help`: Show a help message.

Example usage:
```bash
# Displaying a derived key
snpguest display key derived-key.bin
```
---

> # SNP Examples

Reviewer: Your examples are just a repeat from what you have above in commands. I would maybe use this section to show how would you go about going through a regular attestation workflow and an extended attestation workflow. This in addition to the flowcharts you added at the bottom would be very helpful.

```

## Verifying Attestation

### Verifying certificate chain:

```bash
snpguest verify certs <certs-directory>
```

- `<certs-directory>`: Specify the path to a directory where the certificates are stored. A full certificate chain is required (ASK, ARK, and VCEK).

Example:

```bash
snpguest verify certs ./certs
```

### Verifying Attestation Report:

```bash
snpguest verify attestation <certs-directory> <attestation-report-path> [-t, --tcb] [-s, --signature]
```

- `<certs-directory>`: Specify the path to a directory where the certificates are stored.
- `<attestation-report-path>`: Specify the path to an attestation report to be verified.
- `-t, --tcb`: Just verify the TCB contents of the attestation report.
- `-s, --signature`: Just verify the signature of the attestation report.

Example:

```bash
# Verify Attestation
snpguest verify attestation ./certs attestation-report.bin
# Verify Attestation Signature only
snpguest verify attestation ./certs attestation-report.bin --signature
```

---
## Extended Attestation Flowchart
![alt text](https://github.com/virtee/snpguest/blob/main/docs/extended.png?raw=true)
## Regular Attestation Flowchart

> ![alt text](https://github.com/virtee/snpguest/blob/main/docs/regular.png?raw=true)

Reviewer: I'm not sure if this is how you attach images. At least for me, in you repository, I can't see the images in the README.

Reviewer 2: I cannot see them, either
